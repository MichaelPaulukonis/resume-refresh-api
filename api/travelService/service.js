import { marked } from 'marked';
import sanitizeHtml from 'sanitize-html';
import slugify from 'slugify';
import logger from '../lib/logger.js'

// --- RENDER FUNCTIONS ---

function renderMetaSection(meta) {
  if (!meta) return '';
  return `<section class="meta-info">
      <h2>Trip Details</h2>
      <p><strong>Destination:</strong> ${escapeHtml(meta.destination)}</p>
      <p><strong>Dates:</strong> ${escapeHtml(meta.dates)}</p>
      <p><strong>Group:</strong> ${escapeHtml(String(meta.groupSize))} people (${escapeHtml(meta.groupDetails)})</p>
      <p class="small-muted">Report generated on ${escapeHtml(meta.generated_at)}</p>
    </section>`;
}

export function renderPrice(item) {
  if (!item) return '';
  const { price_value, currency, price_extraction_note } = item;

  let priceHtml = '';
  const currencySymbol = currency === 'USD' ? '$' : '';
  const formattedCurrency = currency || '';

  if (typeof price_value === 'string' && price_value.trim()) {
    priceHtml = `<div class="price">${escapeHtml(price_value)} ${formattedCurrency}</div>`;
  }

  const noteHtml = price_extraction_note ? `<div class="price-note">${escapeHtml(price_extraction_note)}</div>` : '';

  return priceHtml + noteHtml;
}

export function renderContactInfo(item) {
  let out = '<div class="contacts">';
  if (item.phone_local) {
    const telLink = item.phone_e164 || item.phone_local.replace(/[^+\d]/g, '');
    out += `<a class="tel" href="tel:${escapeHtml(telLink)}">${escapeHtml(item.phone_local)}</a>`;
  }
  if (item.email) {
    out += ` <a class="email" href="mailto:${escapeHtml(item.email)}">${escapeHtml(item.email)}</a>`;
  }
  if (item.website) {
    const href = item.website.startsWith('http') ? item.website : 'https://' + item.website;
    out += ` <a class="website" href="${escapeHtml(href)}" target="_blank" rel="noopener">${escapeHtml(item.website)}</a>`;
  }
  out += '</div>';
  return out;
}

export function renderItemFooter(item) {
  return `<div class="foot small-muted">Confidence: ${escapeHtml(String(item.confidence || 'n/a'))} ${item.source_urls ? renderSourceLinks(item.source_urls) : ''}</div>`;
}

export function renderItem(it) {
  const addressHtml = it.address ? `<div class="address">${escapeHtml(it.address)}</div>` : '';
  return `<article class="item card" id="${slugify(it.id || it.name || 'item', { lower: true, strict: true })}">
        <h3>${escapeHtml(it.name || 'Untitled')}</h3>
        <p>${escapeHtml(it.short_description || '')}</p>
        ${addressHtml}
        ${renderPrice(it)}
        ${renderContactInfo(it)}
        ${renderItemFooter(it)}
      </article>`;
}

function renderCategorySection(categoryName, items) {
  logger.info('category: %j', categoryName);
  logger.info('items: %j', items)
  if (!Array.isArray(items) || items.length === 0) return '';
  let out = `<section class="category"><h2>${escapeHtml(pretty(categoryName))}</h2><div class="items">`;
  for (const it of items) {
    out += renderItem(it);
  }
  out += `</div></section>`;
  return out;
}

function renderContactsSection(vcards) {
  if (!vcards || vcards.length === 0) return '';
  let out = `<section class="top-contacts"><h2>Contacts</h2>`;
  vcards.forEach(c => {
    const slug = slugify(c.name || 'contact', { lower: true, strict: true });
    const tel = c.tel ? c.tel.replace(/[^+\d]/g, '') : '';
    const websiteHtml = c.website ? `<div class="website"><a href="${escapeHtml(c.website)}" target="_blank" rel="noopener">${escapeHtml(c.website)}</a></div>` : '';
    const noteHtml = c.note ? `<div class="note">${escapeHtml(c.note)}</div>` : '';
    out += `<div class="contact-card">
        <div class="name">${escapeHtml(c.name)}</div>
        <div class="tel">${c.tel ? `<a href="tel:${escapeHtml(tel)}">${escapeHtml(c.tel)}</a>` : 'â€”'}</div>
        ${c.email ? `<div class="email"><a href="mailto:${escapeHtml(c.email)}">${escapeHtml(c.email)}</a></div>` : ''}
        ${websiteHtml}
        ${noteHtml}
      </div>`;
  });
  out += `</section>`;
  return out;
}

function renderSourcesSection(sources) {
  if (!sources || sources.length === 0) return '';
  let out = `<section class="sources"><h2>Data Sources</h2><ul>`;
  sources.forEach(source => {
    const href = source.startsWith('http') ? source : 'https://' + source;
    out += `<li><a href="${escapeHtml(href)}" target="_blank" rel="noopener">${escapeHtml(source)}</a></li>`;
  });
  out += `</ul></section>`;
  return out;
}

function renderParsedJsonAsHtml(parsed) {
  if (!parsed) return '';
  const categories = parsed.categories || {};
  let out = '<div class="travel-json">';

  for (const key of Object.keys(categories)) {
    out += renderCategorySection(key, categories[key]);
  }

  const contacts = parsed.contacts_vcards || parsed.top_contacts_vcards;
  out += renderContactsSection(contacts);

  out += '</div>';
  return out;
}

// --- HELPERS ---

function renderSourceLinks(urls) {
  if (!urls) return '';
  return urls.slice(0, 2).map(u => {
    const href = u.startsWith('http') ? u : 'https://' + u;
    return `<a href="${escapeHtml(href)}" target="_blank" rel="noopener">source</a>`;
  }).join(' ');
}

function escapeHtml(s) {
  if (!s) return '';
  const map = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#39;'
  };
  return String(s).replace(/[&<>"'\/]/g, c => map[c] || c);
}

function pretty(s) {
  return s.replace(/_/g, ' ').replace(/\b\w/g, c => c.toUpperCase());
}

// --- MAIN EXPORT ---

export const parseResponse = (travelPlan) => {
  // The travelPlan is now a direct JSON object.
  const metaHtml = renderMetaSection(travelPlan.meta);
  const reportHtml = travelPlan.report ? marked(travelPlan.report) : '';

  // Render the structured categories and contacts from the JSON object
  const jsonHtml = renderParsedJsonAsHtml(travelPlan);
  const sourcesHtml = renderSourcesSection(travelPlan.meta?.sources);

  // Combine and sanitize
  const combined = `<div class="travel-output" id="travelOutput">${metaHtml}${reportHtml}${jsonHtml}${sourcesHtml}</div>`;
  logger.info('combined: %j', combined);

  const safe = sanitizeHtml(combined, {
    allowedTags: sanitizeHtml.defaults.allowedTags.concat(['section', 'article', 'h2', 'h3', 'div']),
    allowedAttributes: {
      a: ['class', 'href', 'target', 'rel'],
      article: ['class', 'id'],
      div: ['class', 'id'],
      section: ['class']
    },
    allowedSchemesByTag: {
      a: ['http', 'https', 'mailto', 'tel']
    }
  });

  const css = `<style>
    .travel-output { font-family: system-ui, -apple-system, "Segoe UI", Roboto, "Helvetica Neue", Arial; color: #0f172a; }
    .travel-output .card { border-radius:6px; padding:12px; margin-bottom:10px; background:#fff; }
    .travel-output .small-muted { color:#6b7280; font-size:0.95rem; }
    .contact-card { margin-bottom: 18px; }
  </style>`;

  return (css + safe);
};