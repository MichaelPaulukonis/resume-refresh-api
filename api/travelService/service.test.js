import { describe, it, expect } from 'vitest';
import { parseResponse, extractFencedJson, safeJsonParse } from './service.js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const sampleLogData = fs.readFileSync(path.join(__dirname, '__fixtures__', 'sampleLog.txt'), 'utf-8');

describe('Travel Service', () => {
  describe('extractFencedJson', () => {
    it('should extract JSON from markdown', () => {
      const { jsonText, md } = extractFencedJson(sampleLogData);
      expect(jsonText).not.toBeNull();
      expect(md).not.toContain('```json');
    });

    it('should return null for jsonText if no JSON is present', () => {
      const { jsonText, md } = extractFencedJson('Just some markdown');
      expect(jsonText).toBeNull();
      expect(md).toBe('Just some markdown');
    });
  });

  describe('safeJsonParse', () => {
    it('should parse valid JSON', () => {
      const { data, error } = safeJsonParse('{"key": "value"}');
      expect(data).toEqual({ key: 'value' });
      expect(error).toBeNull();
    });

    it('should handle trailing commas', () => {
      const { data, error } = safeJsonParse('{"key": "value",}');
      expect(data).toEqual({ key: 'value' });
      expect(error).toBeNull();
    });

    it('should return an error for invalid JSON', () => {
      const { data, error } = safeJsonParse('{"key": "value}');
      expect(data).toBeNull();
      expect(error).not.toBeNull();
    });
  });

  describe('parseResponse', () => {
    it('should generate a snapshot of the HTML output', () => {
      const html = parseResponse(sampleLogData);
      expect(html).toMatchSnapshot();
    });
  });
});