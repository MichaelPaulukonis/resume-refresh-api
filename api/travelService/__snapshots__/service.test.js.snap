// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Travel Service > parseResponse (complex) > should generate a snapshot of the HTML output 1`] = `
"<style>
      .travel-output { font-family: system-ui, -apple-system, "Segoe UI", Roboto, "Helvetica Neue", Arial; color: #0f172a;
  }
      .travel-output .card { border-radius:6px; padding:12px; margin-bottom:10px; background:#fff; }
      .travel-output .small-muted { color:#6b7280; font-size:0.95rem; }
    </style><div class="travel-output"><p>Here is a curated travel program in JSON format, followed by a human-friendly summary with contact cards.</p>
<p>JSON Output:</p>
<p>Human Summary:</p>
<ul>
<li>Flexible options for women 65+ visiting NYC December 19–27, 2025, combining iconic sights, markets, cultural highlights, outdoor escapes, and day trips.</li>
<li>Expect cold weather, possibly snow—dress in warm layers.</li>
<li>Cultural &amp; Historical Highlight:<ul>
<li>Midnight Mass at St. Patrick’s Cathedral on Christmas Eve—free, deeply reflective, arrive early.</li>
</ul>
</li>
<li>Markets &amp; Shopping:<ul>
<li>Bryant Park Winter Village (free entry, skate rental ~USD 20)</li>
<li>Union Square and Columbus Circle Holiday Markets—free entry, great for browsing artisan gifts.</li>
<li>Grand Central Holiday Fair—indoor, warm, ideal for leisurely shopping.</li>
</ul>
</li>
<li>Seasonal Highlights:<ul>
<li>Rockefeller Center Christmas Tree &amp; skating—iconic, festive.</li>
<li>Radio City Christmas Spectacular—classic show (tickets USD 50–150).</li>
<li>LuminoCity Festival—immersive lights (ticket ~USD 30–40).</li>
</ul>
</li>
<li>Outdoor &amp; Nature:<ul>
<li>Bear Mountain State Park—ice skating, winter hiking, close to the city.</li>
</ul>
</li>
<li>Day Trips:<ul>
<li>Cold Spring—quaint Hudson River town, antiques, cafés.</li>
<li>Beacon—with galleries and Dia Beacon museum.</li>
<li>Mohonk Mountain House—spa, skating, hiking, meals in one.</li>
</ul>
</li>
<li>Evening &amp; Special Event:<ul>
<li>Knicks vs. Cavaliers at MSG on Christmas Day—festive sports outing (tickets via secondary market).</li>
</ul>
</li>
<li>Top 3 Contact Cards:</li>
</ul>
<ol>
<li><p>Rockefeller Center</p>
<ul>
<li>Phone: +1 (212) 625‑0000</li>
<li>Email: <a href="mailto:info@rockefellerplaza.com">info@rockefellerplaza.com</a></li>
</ul>
</li>
<li><p>St. Patrick’s Cathedral</p>
<ul>
<li>Phone: +1 (212) 753‑5000</li>
<li>Email: <a href="mailto:info@saintpatrickscathedral.org">info@saintpatrickscathedral.org</a></li>
</ul>
</li>
<li><p>Madison Square Garden (Knicks game)</p>
<ul>
<li>Phone: +1 (212) 548‑1111</li>
<li>Website: msg.com</li>
</ul>
</li>
</ol>
<div class="travel-json"><section class="category"><h2>Accommodations</h2><div class="items"><article class="item card">
        <h3>Rockefeller Plaza Hotels (e.g., The Plaza, Rockefeller Center area)</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">approx. USD 400–600</div><div class="price-note">Typical rates for mid-December from hotel booking platforms</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.5 </div>
      </article></div></section><section class="category"><h2>Cultural &amp; Historical Highlights</h2><div class="items"><article class="item card">
        <h3>St. Patrick’s Cathedral Midnight Mass</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: n/a </div>
      </article></div></section><section class="category"><h2>Local Food &amp; Dining</h2><div class="items"><article class="item card">
        <h3>Beacon cafés and eateries (Beacon, NY day trip)</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">approx. USD 20–40</div><div class="price-note">Estimated from CTraveler recommendations of casual dining</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.4 </div>
      </article></div></section><section class="category"><h2>Markets &amp; Shopping</h2><div class="items"><article class="item card">
        <h3>Bryant Park Winter Village &amp; Holiday Shops</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Free admission; skate rental estimated USD 20</div><div class="price-note">From event listing stating free admission, typical rental fee</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.7 </div>
      </article><article class="item card">
        <h3>Union Square Holiday Market</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Free entry; purchases vary</div><div class="price-note">From market webpage listing free entry</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.9 </div>
      </article><article class="item card">
        <h3>Grand Central Holiday Fair</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Free entry; purchases vary</div><div class="price-note">From Christmasmarketusa.com listing</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.9 </div>
      </article></div></section><section class="category"><h2>Outdoor Adventures &amp; Nature</h2><div class="items"><article class="item card">
        <h3>Bear Mountain State Park (ice skating, hiking)</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Parking: USD 8 weekdays / USD 10 weekends</div><div class="price-note">From RadicalStorage and MightyTravels articles</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.8 </div>
      </article></div></section><section class="category"><h2>Day Trips / Excursions</h2><div class="items"><article class="item card">
        <h3>Cold Spring, NY</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Train fare approx. USD 15–20 round-trip</div><div class="price-note">Estimated Metro-North fare</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.6 </div>
      </article><article class="item card">
        <h3>Beacon, NY (Dia Beacon)</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">USD 25 (museum admission)</div><div class="price-note">Typical Dia Beacon admission</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.7 </div>
      </article><article class="item card">
        <h3>Mohonk Mountain House Day Pass</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">approx. USD 150–200</div><div class="price-note">Estimated day-pass from DominicanAbroad tour info</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.6 </div>
      </article></div></section><section class="category"><h2>Evening &amp; Nightlife Experiences</h2><div class="items"><article class="item card">
        <h3>Knicks vs Cavaliers at Madison Square Garden (Christmas Day)</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Secondary market pricing varies (likely USD 100+)</div><div class="price-note">From New York Post article about ticket sales</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.8 </div>
      </article></div></section><section class="category"><h2>Seasonal Or Special Events</h2><div class="items"><article class="item card">
        <h3>Rockefeller Center Christmas Tree &amp; Ice Skating</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Free to view; skating rental cost typical USD 20</div><div class="price-note">From TheChristmas.com listing for event</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.9 </div>
      </article><article class="item card">
        <h3>Radio City Christmas Spectacular starring the Rockettes</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Tickets typically USD 50–150</div><div class="price-note">From TheChristmas.com listing general pricing</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.8 </div>
      </article><article class="item card">
        <h3>Bryant Park Winter Village</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Free entry; skate rental approx. USD 20</div><div class="price-note">From event listing</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.7 </div>
      </article><article class="item card">
        <h3>LuminoCity Festival</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">Estimated USD 30–40</div><div class="price-note">Typical ticket price for light festivals</div>
        <div class="contacts"></div>
        <div class="foot small-muted">Confidence: 0.5 </div>
      </article></div></section></div></div>"
`;

exports[`Travel Service > parseResponse (simple) > should generate a snapshot of the HTML output 1`] = `
"<style>
      .travel-output { font-family: system-ui, -apple-system, "Segoe UI", Roboto, "Helvetica Neue", Arial; color: #0f172a;
  }
      .travel-output .card { border-radius:6px; padding:12px; margin-bottom:10px; background:#fff; }
      .travel-output .small-muted { color:#6b7280; font-size:0.95rem; }
    </style><div class="travel-output"><p>Here is a curated, flexible travel program for your group of nine women (65+) visiting New York City from 2025‑12‑20 to 2025‑12‑27. The offerings span cultural highlights, markets, food, outdoor experiences, and more. Each item includes contact details, pricing, source citations, and notes on holiday-specific scheduling or potential closures.</p>
<div class="travel-json"><section class="category"><h2>Accommodations</h2><div class="items"><article class="item card">
        <h3>Bryant Park Winter Village area hotels</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">varies, approx. USD 300–500</div><div class="price-note">estimated from nearby hotel rates in Midtown</div>
        <div class="contacts"><a class="tel" href="tel:+121276842422127684242">+1‑212‑768‑4242 (local: 212‑768‑4242)</a> <a class="website" href="http://www.bryantpark.org" target="_blank" rel="noopener">http://www.bryantpark.org</a></div>
        <div class="foot small-muted">Confidence: 0.6 </div>
      </article></div></section><section class="category"><h2>Cultural &amp; Historical Highlights</h2><div class="items"><article class="item card">
        <h3>Radio City Christmas Spectacular (Rockettes)</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">approx. USD 80–150</div><div class="price-note">typical Broadway ticket range from official site / ticket sellers</div>
        <div class="contacts"><a class="tel" href="tel:+121223962002122396200">+1‑212‑239‑6200 (local: 212‑239‑6200)</a> <a class="website" href="https://www.rockettes.com/christmas/" target="_blank" rel="noopener">https://www.rockettes.com/christmas/</a></div>
        <div class="foot small-muted">Confidence: 0.7 </div>
      </article><article class="item card">
        <h3>The Nutcracker at Lincoln Center</h3>
        <div class="meta small-muted"> </div>
        <p></p>
        <div class="price">approx. USD 70–200</div><div class="price-note">based on previous season pricing</div>
        <div class="contacts"><a class="tel" href="tel:+121287055702128705570">+1‑212‑870‑5570 (local: 212‑870‑5570)</a> <a class="website" href="https://www.nycballet.com" target="_blank" rel="noopener">https://www.nycballet.com</a></div>
        <div class="foot small-muted">Confidence: 0.6 </div>
      </article></div></section></div></div>"
`;
